ops/dataset-copy.sh

```diff
@@ -0,0 +1,177 @@
+#!/bin/bash
+# a script to copy a dataset from one deployment to another
+
+set -eu
+
+### INPUTS ###
+DATASET_UUID_TO_COPY="00000000-0000-0000-0000-000000000000"
+AUTH_TOKEN="authorization: Bearer ya29"
+
+SRC_URL="https://jade.datarepo-dev.broadinstitute.org"
+DST_URL="https://jade-fb.datarepo-dev.broadinstitute.org/"
+
+### utility functions ###
+token_check() {
+    RESULT=$(curl -s -X GET "${SRC_URL}/api/resources/v1/profiles" -H "accept: application/json" -H "${AUTH_TOKEN}")
+    if [[ ${RESULT} =~ "<title>401 Unauthorized</title>" ]]; then
+        echo "Error: AUTH_TOKEN is invalid or expired"
+        exit 1
+    fi
+}
+
+uuid_lc() {
+    uuidgen | tr '[:upper:]' '[:lower:]'
+}
+
+post_url() {
+    curl -s -X POST "${1}" -H  "accept: application/json" -H  "Content-Type: application/json" -H "${AUTH_TOKEN}" -d "${2}"
+}
+
+get_url() {
+    curl -s -X GET  "${1}" -H  "accept: application/json" -H  "Content-Type: application/json" -H "${AUTH_TOKEN}"
+}
+
+CURRENT_DATE=$(date +%Y%m%d)
+
+### PART 1: create profile ###
+profile_create() {
+    echo "1: creating new profile"
+
+    PROFILE_URL="${DST_URL}/api/resources/v1/profiles"
+
+    PROFILE_UUID=$(uuid_lc)
+    PROFILE_NAME="profile-search-api-${CURRENT_DATE}-${PROFILE_UUID:0:8}"
+
+    PROFILE_JSON=$(cat << EOF
+{
+  "biller": "direct",
+  "billingAccountId": "00708C-45D19D-27AAFA",
+  "description": "profile created by datasetcopy.sh",
+  "id": "${PROFILE_UUID}",
+  "profileName": "${PROFILE_NAME}"
+}
+EOF
+)
+
+    post_url "${PROFILE_URL}" "${PROFILE_JSON}"
+
+    printf "\n1: created profile %s\n" "${PROFILE_UUID}"
+}
+
+### PART 2: copy dataset schema ###
+dataset_copy_schema() {
+    echo "2: copying dataset schema"
+
+    DATASET_URL="${SRC_URL}/api/repository/v1/datasets/${DATASET_UUID_TO_COPY}?include=SCHEMA&include=ACCESS_INFORMATION"
+
+    DATASET_JSON=$(get_url "${DATASET_URL}")
+    DATASET_NAME=$(echo "${DATASET_JSON}" | jq -r '.name')
+    DATASET_SCHEMA=$(echo "${DATASET_JSON}" \
+        | jq 'del(.id, .dataProject, .defaultSnapshotId, .createdDate, .storage, .accessInformation) | . + { "cloudPlatform": "gcp", "region": "US" } | .defaultProfileId = "'"${PROFILE_UUID}"'"' \
```

Comment:
What if the original dataset isn't in `US`?

Reply:
It will copy the dataset into a US region, regardless of the dataset origin. I had to do this because the GCS bucket, BQ, and instances all need to be in the same region for this to work, and this was the simplest approach. But I agree this is not optimal in the general case, say if you wanted to copy a dataset from one non-US region to another. Since this is mainly for testing purposes and moving datasets from one dev deployment to another, I didn't bother making it configurable. To make it configurable would probably not be trivial, since you would need to ensure all the resources are all created in the same region.

Reply:
ay that does sound like a lot of work.